### PR TITLE
Onboarding redirection on mobile

### PR DESCRIPTION
**Pull request by maximemenager**

### CODE REVIEW

File: web/auth/auth.go [hunk 1]
```diff
@@ -1122,4 +1149,5 @@ func Routes(router *echo.Group) {
 	authorizeGroup.POST("/app", authorizeApp)
 
 	router.POST("/access_token", accessToken)
+	router.POST("/secret_exchange", secretExchange)
```
Comment (nono) [inline web/auth/auth.go]: Note for later: this route should be documented in `docs/auth.md`

File: pkg/instance/instance.go [hunk 2]
```diff
@@ -98,9 +98,14 @@ type Instance struct {
...
 	Dev           bool     `json:"dev,omitempty"`            // Whether or not the instance is for development
 
-	OnboardingFinished bool  `json:"onboarding_finished,omitempty"` // Whether or not the onboarding is complete.
-	BytesDiskQuota     int64 `json:"disk_quota,string,omitempty"`   // The total size in bytes allowed to the user
-	IndexViewsVersion  int   `json:"indexes_version"`
+	OnboardingFinished bool `json:"onboarding_finished,omitempty"` // Whether or not the onboarding is complete.
+
+	OnboardingSecret      string `json:"onboarding_secret,omitempty"`
+	OnboardingApp         string `json:"onboarding_app,omitempty"`
+	OnboardingPermissions string `json:"onboarding_permissions,omitempty"`
```
Comment (nono) [inline pkg/instance/instance.go]: I'd have put these 3 fields in the oauth client, not in the instance.

File: web/auth/auth.go [hunk 3]
```diff
@@ -1086,6 +1087,32 @@ func passphraseRenew(c echo.Context) error {
...
+
+	doc, err := oauth.FindClient(instance, instance.OnboardingClientID)
+
+	if err == nil {
+		if instance.OnboardingSecret != "" && instance.OnboardingSecret == e.Secret {
+			return c.JSON(http.StatusOK, doc)
+		}
+		r = "Bad secret"
+	}
+	return jsonapi.Errorf(http.StatusBadRequest, "%s", r)
```
Comment (nono) [inline web/auth/auth.go]: I prefer to handle the error first, and to use the correct HTTP codes:

```go
doc, err := oauth.FindClient(instance, instance.OnboardingClientID)
if err != nil {
    return jsonapi.NotFound(err)
}
if instance.OnboardingSecret == "" || instance.OnboardingSecret != e.Secret {
    return jsonapi.InvalidAttribute("secret", errors.New("Invalid secret"))
}
return c.JSON(http.StatusOK, doc)
}

File: web/auth/auth.go [hunk 4]
```diff
@@ -1086,6 +1087,32 @@ func passphraseRenew(c echo.Context) error {
 	return c.Redirect(http.StatusSeeOther, inst.PageURL("/auth/login", nil))
 }
 
+// Used to trade a secret for OAuth client informations
+func secretExchange(c echo.Context) error {
+	type exchange struct {
+		Secret string `json:"secret,omitempty"`
```
Comment (nono) [inline web/auth/auth.go]: Why `omitempty` here?

File: web/auth/auth.go [hunk 5]
```diff
@@ -1086,6 +1087,32 @@ func passphraseRenew(c echo.Context) error {
 	return c.Redirect(http.StatusSeeOther, inst.PageURL("/auth/login", nil))
 }
 
+// Used to trade a secret for OAuth client informations
+func secretExchange(c echo.Context) error {
```
Comment (nono) [inline web/auth/auth.go]: Unit tests are expected for this function

File: web/settings/context.go [hunk 6]
```diff
@@ -47,6 +54,40 @@ func onboarded(c echo.Context) error {
...
+		deeplink := fmt.Sprintf("cozy%s://%s", i.OnboardingApp, i.Domain)
+
+		// Generate the OAuth client
+		OAuthClient := oauth.Client{
+			RedirectURIs: []string{deeplink},
+			ClientName:   i.OnboardingApp,
+			SoftwareID:   "github.com/cozy/cozy-stack",
+		}
+		OAuthClient.Create(i)
+		i.OnboardingClientID = OAuthClient.ClientID
```
Comment (nono) [inline web/settings/context.go]: In our sequence diagram, it's the cloudery that calls the stack to create the OAuth client

File: web/settings/context.go [hunk 7]
```diff
@@ -47,6 +54,40 @@ func onboarded(c echo.Context) error {
...
+
+		// Redirection
+		// Generate state
+		b := make([]byte, 32)
```
Comment (nono) [inline web/settings/context.go]: I've updated the sequence diagram: it should be the webview that creates this state. And the state should be checked by the webview after the redirection (ie the state the webview has generated == the state in the query string in the redirected URL)

File: web/auth/auth.go [hunk 8]
```diff
@@ -1128,6 +1130,36 @@ func passphraseRenew(c echo.Context) error {
...
+
+	if err != nil {
+		return jsonapi.NotFound(err)
+	}
+
+	if doc.OnboardingSecret == "" || doc.OnboardingSecret != e.Secret {
+		return jsonapi.InvalidAttribute("secret", errors.New("Invalid secret"))
+	}
+
+	return c.JSON(http.StatusOK, doc)
```
Comment (nono) [inline web/auth/auth.go]: You should use the same format for the response that the `GET /auth/register/:client_id`: https://docs.cozy.io/en/cozy-stack/auth/#get-authregisterclient-id

File: web/settings/context.go [hunk 9]
```diff
@@ -47,6 +51,34 @@ func onboarded(c echo.Context) error {
 		}
 	}
 	redirect := i.OnboardedRedirection().String()
+
+	// Retreiving client
+	clientID := c.QueryParam("client_id")
```
Comment (nono) [inline web/settings/context.go]: I don't think we have a `client_id` in the query-string here!

File: web/settings/context.go [hunk 10]
```diff
@@ -47,6 +51,34 @@ func onboarded(c echo.Context) error {
...
+	client, err := oauth.FindClient(i, clientID)
+	if err != nil {
+		return err
```
Comment (nono) [inline web/settings/context.go]: No, on the web onboarding, we have no client, but we should not `return err` here!

Comment (nono) [reply web/settings/context.go]: And this shows the need of unit tests ;-)

File: pkg/oauth/client.go [hunk 11]
```diff
@@ -70,6 +70,11 @@ type Client struct {
 
 	// XXX omitempty does not work for time.Time, thus the interface{} type
 	SynchronizedAt interface{} `json:"synchronized_at,omitempty"` // Date of the last synchronization, updated by /settings/synchronized
+
+	OnboardingSecret      string `json:"onboarding_secret,omitempty"`
+	OnboardingApp         string `json:"onboarding_app,omitempty"`
+	OnboardingPermissions string `json:"onboarding_permissions,omitempty"`
+	OnboardingState       string `json:"onboarding_state,omitempty"`
```
Comment (nono) [inline pkg/oauth/client.go]: How these fields will be filled? We need a CLI command to do that, no?

Comment (nono) [reply pkg/oauth/client.go]: Note: it can be done in another PR if you don't want to work too long on this one

File: web/settings/context.go [hunk 12]
```diff
@@ -47,6 +50,27 @@ func onboarded(c echo.Context) error {
...
+		queryParams := url.Values{
+			"client_id":     {client.ClientID},
+			"redirect_uri":  {deeplink},
+			"state":         {client.OnboardingState},
+			"response_type": {"code"},
+			"scope":         {client.OnboardingPermissions},
+		}
+		redirect = i.PageURL("/auth/authorize", queryParams)
+
+	}
```
Comment (nono) [inline web/settings/context.go]: I would also like a unit test for this

File: web/auth/auth.go (hunk 1 above)
Comment (edas) [reply web/auth/auth.go]: done

**Pull request merged**